src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp

```diff
@@ -48,22 +48,45 @@ class CgroupV2Controller: public CgroupController {
     char *subsystem_path() { return _path; }
 };
 
+class CgroupV2CpuController: public CgroupV2Controller, public CgroupCpuController {
+  public:
+    CgroupV2CpuController(char * mount_path, char *cgroup_path) : CgroupV2Controller(mount_path, cgroup_path) {
+    }
+    int cpu_quota();
+    int cpu_period();
+    int cpu_shares();
+};
+
+class CgroupV2MemoryController: public CgroupV2Controller, public CgroupMemoryController {
+  public:
+    CgroupV2MemoryController(char * mount_path, char *cgroup_path) : CgroupV2Controller(mount_path, cgroup_path) {
+    }
+
+    jlong read_memory_limit_in_bytes(julong upper_bound);
+    jlong memory_and_swap_limit_in_bytes(julong host_mem, julong host_swp);
+    jlong memory_and_swap_usage_in_bytes(julong host_mem, julong host_swp);
+    jlong memory_soft_limit_in_bytes(julong upper_bound);
+    jlong memory_usage_in_bytes();
+    jlong memory_max_usage_in_bytes();
+    jlong rss_usage_in_bytes();
+    jlong cache_usage_in_bytes();
+    void print_version_specific_info(outputStream* st, julong host_mem);
+};
+
 class CgroupV2Subsystem: public CgroupSubsystem {
   private:
     /* One unified controller */
-    CgroupController* _unified = nullptr;
+    CgroupV2MemoryController* _unified = nullptr;
     /* Caching wrappers for cpu/memory metrics */
-    CachingCgroupController* _memory = nullptr;
-    CachingCgroupController* _cpu = nullptr;
-
-    jlong mem_swp_limit_val();
-    jlong mem_swp_current_val();
+    CachingCgroupController<CgroupMemoryController*>* _memory = nullptr;
+    CachingCgroupController<CgroupCpuController*>* _cpu = nullptr;
 
   public:
-    CgroupV2Subsystem(CgroupController * unified) {
-      _unified = unified;
-      _memory = new CachingCgroupController(unified);
-      _cpu = new CachingCgroupController(unified);
+    CgroupV2Subsystem(CgroupV2MemoryController * memory,
+                      CgroupV2CpuController* cpu) {
+      _unified = memory; // Use memory for now, should have all separate later
+      _memory = new CachingCgroupController<CgroupMemoryController*>(memory);
+      _cpu = new CachingCgroupController<CgroupCpuController*>(cpu);
```

Comment:
Style: hug the asterisk to the name

Pre-existing: Please initialize the members in intiializer list.